packages/docutils/lib/builder/nav.ts

```diff
@@ -0,0 +1,320 @@
+ *
+ * @module
+ */
+
+import {fs} from '@appium/support';
+import _ from 'lodash';
+import path from 'node:path';
+import {
+  DEFAULT_NAV_HEADER,
+  DEFAULT_REL_TYPEDOC_OUT_PATH,
+  NAME_BIN,
+  NAME_MKDOCS_YML,
+  NAME_TYPEDOC_JSON,
+} from '../constants';
+import {DocutilsError} from '../error';
+import {
+  findDirsIn,
+  findMkDocsYml,
+  findTypeDocJsonPath,
+  readTypedocJson,
+  readYaml,
+  safeWriteFile,
+  stringifyYaml,
+} from '../fs';
+import logger from '../logger';
+import {MkDocsYml, MkDocsYmlNav} from '../model';
+import {relative} from '../util';
+
+const log = logger.withTag('builder:nav');
+
+/**
+ * Gets a list of `.md` files relative to `docs_dir`
+ * @param targetDir Directory ostensibly containing Markdown files; must be absolute
+ * @param mkDocsDocsDir The path to the `docs_dir` in the `mkdocs.yml` file; must be absolute
+ * @returns List of Markdown files relative to the `docs_dir` in the `mkdocs.yml` file
+ */
+async function findRelativeMarkdownFiles(
+  targetDir: string,
+  mkDocsDocsDir: string
+): Promise<string[]> {
+  if (!path.isAbsolute(targetDir)) {
+    throw new DocutilsError(`Expected absolute path, got '${targetDir}'`);
+  }
+  if (!path.isAbsolute(mkDocsDocsDir)) {
+    throw new DocutilsError(`Expected absolute path, got '${mkDocsDocsDir}'`);
+  }
+  const relDir = path.relative(mkDocsDocsDir, targetDir);
+  const dirEnts = await fs.readdir(targetDir, {withFileTypes: true});
+  return dirEnts
+    .filter((ent) => ent.isFile() && ent.name.endsWith('.md'))
+    .map((ent) => path.join(relDir, ent.name));
+}
+
+/**
+ * Because the `nav` property of `mkdocs.yml` is both a recursive type and a kind of awful one, it's
+ * easier to work with it if we rewrite the data into a flat array of objects. We keep a `keypath`
+ * prop which represents the deep/nested location within the `nav` object.
+ *
+ * @privateRemarks This function is not recursive; instead it loops over a queue of items to process
+ * data, and we append to that queue while processing if needed.
+ * @param nav Contents of the `nav` prop of `mkdocs.yml`
+ * @returns A list of objects, each with a `keypath` property and a `fileOrUrl` property (and maybe
+ * a `name` property)
+ */
+export function parseNav(nav: MkDocsYmlNav): ParsedNavData[] {
+  let parsedNav: ParsedNavData[] = [];
+  const entries = Object.entries(nav);
+  type QueueItem = {
+    entries: typeof entries;
+    keypath: string;
+  };
+  const queue: QueueItem[] = [{entries, keypath: ''}];
+
+  while (queue.length) {
+    const {entries, keypath} = queue.shift()!;
+    for (const [key, item] of entries) {
+      if (_.isString(item)) {
+        const navData: ParsedNavData = {
+          keypath: keypath ? `${keypath}.${key}` : key,
+          fileOrUrl: item,
+        };
+
+        // if the key is not convertible to a number, it's a name
+        // which was manually put there by somebody.
+        if (isNaN(Number(key))) {
```

Comment:
can we use Number.isNaN?

Reply:
why?

Reply:
https://www.w3schools.com/jsref/jsref_isnan_number.asp

isNaN does unnecessary implicit conversion while the latter does not

Reply:
fair enough